### PR TITLE
ai: Add a node embedding watcher

### PR DESCRIPTION
**Pull request by hugoShaka**
This PR adds the embedding core logic in auth:
- add Embeddings service and its local implementation
- add Embedding type and proto message
- add nodeEmbeddingCollector tracking nodes
- add NodeEmbeddingWatcher watching for events adn sending them to the
  collector
- add the Embedder interface and its openai implementation

Fixes https://github.com/gravitational/teleport.e/issues/1464

### CODE REVIEW

**Review by xacrimon — COMMENTED** [review]
first pass

File: lib/services/local/embeddings_test.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
+package local
```
Comment (xacrimon) [inline lib/services/local/embeddings_test.go:1]: missing license notice?

File: lib/services/local/embeddings.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
+package local
```
Comment (xacrimon) [inline lib/services/local/embeddings.go:1]: missing license notice?

File: lib/services/embeddings_test.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
+package services_test
```
Comment (xacrimon) [inline lib/services/embeddings_test.go:1]: missing license notice?

File: lib/services/embeddings.go [hunk 4]
```diff
@@ -0,0 +1,312 @@
+package services
```
Comment (xacrimon) [inline lib/services/embeddings.go:1]: missing license notice?

File: lib/services/embeddings.go [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+			delete(n.currentNodes, nodeName)
+		}
+	}
+}
+
+// RunIndexation walks through all collector-tracked nodes and runs a batch
+// embedding on all nodes needing embeddings. The embeddings are then inserted
+// into the vector index. This process is ran asynchronously to reduce the load
+// and leverage OpenAI's batch embedding API.
+func (n *nodeEmbeddingCollector) RunIndexation(ctx context.Context) error {
```
Comment (jakule) [inline lib/services/embeddings.go:188]: If I understand this correctly, if two goroutines run this method, both will read all nodes, then both will create embeddings and update the index with the same values. Shouldn't we just take the exclusive lock at the beginning and double-check if the embedding generation is needed to avoid it?

File: lib/services/embeddings_test.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+		vector := make([]float32, len(hash))
+		for j, x := range hash {
+			vector[j] = 1 / float32(int(x)+1)
+		}
+		result[i] = vector
+	}
+	return result, nil
+}
+
+func TestNodeEmbeddingWatcher(t *testing.T) {
```
Comment (jakule) [inline lib/services/embeddings_test.go:40]: This test tests a lot of stuff. Can we split it into a few smaller/simpler tests? Otherwise, if something is broken, you don't have clear information what exactly happened.

File: lib/ai/embedding.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+	// ComputeEmbeddings computes the embeddings of multiple strings.
+	// The embedding list follows the input order (e.g. result[i] is the
+	// embedding of input[i]).
+	ComputeEmbeddings(ctx context.Context, input []string) ([][]float32, error)
+}
+
+// ComputeEmbeddings taxes a map of nodes and calls openAI to generate
+// embeddings for those nodes. ComputeEmbeddings is responsible for
+// implementing a retry mechanism if the embedding computation is flaky.
+func (client *Client) ComputeEmbeddings(ctx context.Context, input []string) ([][]float32, error) {
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply lib/ai/embedding.go]: I introduced two types: `Vector32` and `Vector64`. I only use Vector32 everywhere, but @jakule's vector index uses float64, so the second type will become handy and we'll be able to write conversion methods.

File: api/proto/teleport/embedding/v1/embedding.proto [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+// Embedding contains a Teleport resource embedding. Embeddings are small semantic
+// representations of larger and more complex data. Embeddings can be compared,
+// the smaller the distance between two vectors, the closer the concepts are.
+// Teleport Assist embeds resources to perform semantic search.
+message Embedding {
+  // Metadata is the embedding metadata.
+  types.Metadata metadata = 1;
+
+  // Version is the embedding resource version.
+  string version = 2;
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply api/proto/teleport/embedding/v1/embedding.proto:32]: Currently, we don't even need a version, but we would if we decide to change the embedding resource. I made this a string to be consistent with the `Version` field of our other resources.

File: lib/services/embeddings.go [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+type NodeEmbeddingWatcherConfig struct {
+	NodeWatcherConfig
+	Embeddings
+}
+
+// nodeEmbeddingCollector accompanies resourceWatcher when monitoring currentNodes.
+// It keeps tracks of which node has been embedded and which node requires embedding.
+// The embedding happens asynchronously as calling the openAI API by batch is
+// much quicker, stable and efficient.
+type nodeEmbeddingCollector struct {
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply lib/services/embeddings.go:130]: I don't think, 3 of them are here in all other collectors. The initialization channel and its fuse are here to please the resourceCollector interface. Same for the `stale` boolean.

The only synchronization structure used by our custom logic is the mutex.

File: lib/services/embeddings.go [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+	NodeEmbeddingWatcherConfig
+
+	// initializationC is used to check whether the initial sync has completed
+	initializationC chan struct{}
+	once            sync.Once
+
+	currentNodes map[string]*embeddedNode
+
+	rw    sync.RWMutex
+	stale atomic.Bool
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply lib/services/embeddings.go:155]: I added some comments and revamped the locking mechanism, can you check if this makes more sense and if there's information lacking?

File: lib/services/embeddings.go (hunk 5 above)
Comment (hugoShaka) [reply lib/services/embeddings.go:244]: The only sensitive part in this is when the full sync runs and the function getResourcesAndUpdateCurrent acquires the lock.

There's a single routine triggering the embeddings so we should not be doing double work, but even if this happens this does not affect consistency because of the timestamp check.

File: lib/ai/embedding.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (xacrimon) [reply lib/ai/embedding.go]: 👍

File: api/proto/teleport/embedding/v1/embedding.proto (hunk 8 above)
[the remark this answers is not in the document]
Comment (xacrimon) [reply api/proto/teleport/embedding/v1/embedding.proto:32]: to me, keeping consistent there sounds good

File: lib/service/service.go [hunk 11]
```diff
@@ -1694,6 +1701,27 @@ func (process *TeleportProcess) initAuthService() error {
 	}
 	authServer.SetLockWatcher(lockWatcher)
 
+	if authServer.Embeddings != nil {
+		log.Infof("Starting embedding watcher")
```
Comment (jakule) [inline lib/service/service.go:1705]:
```suggestion
		log.Debugf("Starting embedding watcher")
```
Let's try not to log in `Info` without a reason.

**Review by jakule — APPROVED** [review]
I'm fine with the code as long as @justinas confirms the OpenAI API key logic.

File: lib/config/configuration.go [hunk 12]
```diff
@@ -697,6 +697,16 @@ func applyAuthConfig(fc *FileConfig, cfg *servicecfg.Config) error {
 		cfg.Auth.Preference.SetDisconnectExpiredCert(fc.Auth.DisconnectExpiredCert.Value)
 	}
 
+	if fc.Auth.Assist != nil && fc.Auth.Assist.OpenAI != nil {
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply lib/config/configuration.go:700]: I think the auth endpoint to query embeddings should return an error, and the proxies should handle it gracefully by bailing out and returning an error message saying embeddings must be enabled. I don't think it adds much value to handle cases where embeddings are disabled on the auth side, users will just have a poor UX.

File: lib/services/embeddings.go [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+// we use in utils.FastMarshall() but with more precision on floats. This
+// is required to reduce the lost precision when storing embeddings in the
+// backend.
+var PreciseJSONConfig = jsoniter.Config{
+	EscapeHTML:                    false,
+	MarshalFloatWith6Digits:       false, // keep precision
+	ObjectFieldMustBeSimpleString: true,  // do not unescape object field
+	SortMapKeys:                   true,
+}.Froze()
+
```
Comment (jakule) [inline lib/services/embeddings.go:61]: Sorry, I missed that earlier. Can we use the standard JSON library? We had many issues with this one in the past.

File: lib/service/servicecfg/auth.go [hunk 14]
```diff
@@ -105,6 +105,10 @@ type AuthConfig struct {
 	// HTTPClientForAWSSTS overwrites the default HTTP client used for making
 	// STS requests. Used in test.
 	HTTPClientForAWSSTS utils.HTTPDoClient
+
+	// AssistAPIKey is the OpenAI API key.
+	// TODO: This key will be moved to a plugin once support for plugins is implemented.
+	AssistAPIKey string
```
[the remark this answers is not in the document]
Comment (justinas) [reply lib/service/servicecfg/auth.go:111]: This isn't persisted into a resource etc. as far as I can tell? Should be fine.

File: lib/ai/embedding.go [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+
+const (
+	maxOpenAIEmbeddingsPerRequest = 1000
+	EmbeddingPeriod               = 5 * time.Minute
+)
+
+// Sha256Hash is the hash of the embedded content. This hash allows to detect if
+// the embedding is still up-to-date or if the content changed and the resource
+// must be re-embedded.
+type Sha256Hash = [32]byte
```
Comment (espadolini) [inline lib/ai/embedding.go:37]: drive-by nitpick

```suggestion
type Sha256Hash = [sha256.Size]byte
```

from `crypto/sha256`

File: lib/ai/embedding.go [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+	"github.com/gravitational/trace"
+	"github.com/sashabaranov/go-openai"
+
+	embeddingpb "github.com/gravitational/teleport/api/gen/proto/go/teleport/embedding/v1"
+	"github.com/gravitational/teleport/api/types"
+)
+
+const (
+	maxOpenAIEmbeddingsPerRequest = 1000
+	EmbeddingPeriod               = 5 * time.Minute
```
Comment (xacrimon) [inline lib/ai/embedding.go:31]: godoc on a public item?

File: lib/services/embeddings.go [hunk 17]
```diff
@@ -0,0 +1,396 @@
...
+	// a new embedding otherwise. The newly computed embedding is stored in
+	// the backend.
+	Embed(ctx context.Context, kind string, resources map[string][]byte) ([]*ai.Embedding, error)
+}
+
+// PreciseJSONConfig is the fast JSON marshall/unmarshall configuration
+// we use in utils.FastMarshall() but with more precision on floats. This
+// is required to reduce the lost precision when storing embeddings in the
+// backend.
+var PreciseJSONConfig = jsoniter.Config{
```
Comment (xacrimon) [inline lib/services/embeddings.go:55]: will be be swapping jsoniter?

**Review by xacrimon — APPROVED** [review]

File: lib/ai/embedding.go [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+	return strings.Join(nameParts[1:], "/")
+}
+
+// NewEmbedding is an Embedding constructor.
+func NewEmbedding(kind, id string, vector Vector64, hash Sha256Hash) *Embedding {
+	return &Embedding{
+		Embedding: &embeddingpb.Embedding{
+			Metadata: &types.Metadata{
+				Name: kind + "/" + id,
+			},
```
Comment (jakule) [inline lib/ai/embedding.go:91]: I was thinking about when I was working with this code. Is there a reason to keep the metadata here instead of just kind and id? Metadata keeps more information such as namespace, labels, etc. That information is never populated, and I find myself getting the empty value multiple times.

File: lib/ai/embedding.go [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+
+// Embedding contains a Teleport resource embedding. Embeddings are small semantic
+// representations of larger and more complex data. Embeddings can be compared,
+// the smaller the distance between two vectors, the closer the concepts are.
+// Teleport Assist embeds resources to perform semantic search.
+// The Embedding is named after the embedded resource id and kind. For example
+// the SSH node "bastion-01" has the embedding "node/bastion-01".
+type Embedding struct {
+	*embeddingpb.Embedding
+}
```
Comment (espadolini) [inline lib/ai/embedding.go:57]: Maybe?

```suggestion
type Embedding *embeddingpb.Embedding
```

File: lib/services/local/embeddings.go [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+		return nil, trace.Wrap(err)
+	}
+	return embedding, nil
+}
+
+// Embed takes a resource textual representation, checks if the resource
+// already has an up-to-date embedding stored in the backend, and computes
+// a new embedding otherwise. The newly computed embedding is stored in
+// the backend.
+func (e EmbeddingsService) Embed(ctx context.Context, kind string, resources map[string][]byte) ([]*ai.Embedding, error) {
```
Comment (espadolini) [inline lib/services/local/embeddings.go:98]: `lib/services/local` doesn't feel like the correct place for this - so far the services have only had to deal with CPU-bound marshaling and unmarshaling and interacting with the backend, not call to some external service.

File: lib/services/local/embeddings.go [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return services.UnmarshalEmbedding(result.Value)
+}
+
+// GetEmbeddings returns all embeddings for a given kind.
+func (e EmbeddingsService) GetEmbeddings(ctx context.Context, kind string) ([]*ai.Embedding, error) {
+	startKey := backend.Key(embeddingsPrefix, kind)
+	result, err := e.GetRange(ctx, startKey, backend.RangeEnd(startKey), backend.NoLimit)
```
Comment (espadolini) [inline lib/services/local/embeddings.go:58]: This will OOM the auth in a large enough cluster when asking for embeddings for `node`, since we intend to have (at least!) one embedding per active node.

**Review by espadolini — DISMISSED** [review]

File: api/proto/teleport/embedding/v1/embedding.proto (hunk 8 above)
[the remark this answers is not in the document]
Comment (hugoShaka) [reply api/proto/teleport/embedding/v1/embedding.proto]: I dropped the version and store directly proto messages in the backend

File: lib/services/local/embeddings.go (hunk 21 above)
Comment (hugoShaka) [reply lib/services/local/embeddings.go]: I switched to a stream. This function is currently called only in tests so it did not required large logic changes. If/when we implement garbage collection we'll be able to use the stream in a safe way.

File: lib/services/local/embeddings.go [hunk 22]
```diff
@@ -52,26 +53,17 @@ func (e EmbeddingsService) GetEmbedding(ctx context.Context, kind, resourceID st
 	return services.UnmarshalEmbedding(result.Value)
 }
 
-// GetEmbeddings returns all embeddings for a given kind.
-func (e EmbeddingsService) GetEmbeddings(ctx context.Context, kind string) ([]*ai.Embedding, error) {
+// GetEmbeddings returns a stream of embeddings for a given kind.
+func (e EmbeddingsService) GetEmbeddings(ctx context.Context, kind string) stream.Stream[*ai.Embedding] {
 	startKey := backend.Key(embeddingsPrefix, kind)
```
Comment (espadolini) [inline lib/services/local/embeddings.go:58]:
```suggestion
	startKey := backend.ExactKey(embeddingsPrefix, kind)
```

File: lib/services/embeddings.go [hunk 23]
```diff
@@ -0,0 +1,383 @@
...
+	// UpsertEmbedding creates or update a single ai.Embedding in the backend.
+	UpsertEmbedding(ctx context.Context, embedding *ai.Embedding) (*ai.Embedding, error)
+	// Embed takes a resource textual representation, checks if the resource
+	// already has an up-to-date embedding stored in the backend, and computes
+	// a new embedding otherwise. The newly computed embedding is stored in
+	// the backend.
+	Embed(ctx context.Context, kind string, resources map[string][]byte) ([]*ai.Embedding, error)
+}
+
+// MarshalEmbedding marshals the ai.Embedding resource to JSON.
```
Comment (espadolini) [inline lib/services/embeddings.go:53]:
```suggestion
// MarshalEmbedding marshals the ai.Embedding resource to binary ProtoBuf.
```

File: lib/services/embeddings.go [hunk 24]
```diff
@@ -0,0 +1,383 @@
...
+func MarshalEmbedding(embedding *ai.Embedding) ([]byte, error) {
+	data, err := proto.Marshal((*embeddingpb.Embedding)(embedding))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return data, nil
+}
+
+// UnmarshalEmbedding unmarshals JSON into an ai.Embedding resource.
```
Comment (espadolini) [inline lib/services/embeddings.go:62]:
```suggestion
// UnmarshalEmbedding unmarshals binary ProtoBuf into an ai.Embedding resource.
```

File: lib/services/local/embeddings.go (hunk 20 above)
Comment (hugoShaka) [reply lib/services/local/embeddings.go]: I moved the logic out of the Embeddings service. It only performs backend queries and marshalling operations.

File: lib/services/embeddings.go [hunk 25]
```diff
@@ -0,0 +1,396 @@
...
+	count := 0
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+	for _, node := range n.currentNodes {
+		if node.needsEmbedding == needsEmbedding {
+			count += 1
+		}
+	}
+	return count
+}
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply lib/services/embeddings.go:349]: We can optimize this function, but I think it would defeat this counter's purpose. Tests need a way to peak into the collector's state to know what is happening (and the tests are living in a separate `services_test` package, a private function is not usable here). Making this counter smarter and more efficient increases the risks of it returning an inaccurate value because of some collector bug.

File: lib/services/embeddings.go [hunk 26]
```diff
@@ -0,0 +1,396 @@
...
+type NodeEmbeddingWatcherConfig struct {
+	NodeWatcherConfig
+	Embeddings
+}
+
+// nodeEmbeddingCollector accompanies resourceWatcher when monitoring currentNodes.
+// It keeps tracks of which node has been embedded and which node requires embedding.
+// The embedding happens asynchronously as calling the openAI API by batch is
+// much quicker, stable and efficient.
+type nodeEmbeddingCollector struct {
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply lib/services/embeddings.go:124]: Is this comment a blocker, or can we address it later when we'll extend embeddings? This PR is somewhat important to unlock the rest of the Assist PRs for the next features and it would be easier for us to redesign this later.

I can extract the embedding logic out of the collector, but this will complexify even more an [already complex locking situation](https://github.com/gravitational/teleport/pull/27204#discussion_r1223328498) as the collector and something else would need to sync to ensure embedding consistency. Do you think of a specific design that would allow consistent asynchronous embedding not triggered in the watcher?

**Review by r0mant — APPROVED** [review]
Bot.
